From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Tue, 16 Nov 2021 01:25:38 +0000
Subject: [PATCH] Shearable Item Frames


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
index 90f1ecd97eef93316f61bf5d761d33c313e1a077..4428dbadb4a770529786e377842fed2f9844be14 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -187,6 +187,11 @@ public class ItemFrame extends HangingEntity {
                 // CraftBukkit end
                 this.dropItem(source.getEntity(), false);
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
+                // Bamboo start - Make visible if was sheared
+                if (this.isInvisible()) {
+                    this.setInvisible(false);
+                }
+                // Bamboo end
             }
 
             return true;
@@ -440,8 +445,17 @@ public class ItemFrame extends HangingEntity {
                     }
                 }
             } else {
-                this.playSound(this.getRotateItemSound(), 1.0F, 1.0F);
-                this.setRotation(this.getRotation() + 1);
+                // Bamboo start - Shearing makes item frame invisible
+                if (itemstack.is(Items.SHEARS) && player.getAbilities().mayBuild && !this.isInvisible() &&
+                    player.getBukkitEntity().hasPermission("bamboo.frames.shear")) {
+                    this.setInvisible(true);
+                } else {
+                    //
+                    this.playSound(this.getRotateItemSound(), 1.0F, 1.0F);
+                    this.setRotation(this.getRotation() + 1);
+                    //
+                }
+                // Bamboo end
             }
 
             return InteractionResult.CONSUME;
