From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Wed, 10 Nov 2021 23:12:11 +0000
Subject: [PATCH] Bamboo Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ccb8909d3da0bb6ccc6d35e8497c3b4b993b9044..7dbbf9a132b482916ee56a9a5e33f3973f9de1e6 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1912,6 +1912,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Bamboo start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBambooConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bamboo end
+
         /**
          * Sends the component to the player
          *
