From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Wed, 10 Nov 2021 23:12:11 +0000
Subject: [PATCH] Bamboo Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 4e26a6d95c1d8468c0b9cea8d49fafdc4f64d4f9..d0137335822dc596f972d5efc2b506a97bd2ba8c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1911,6 +1911,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Bamboo start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBambooConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bamboo end
+
         /**
          * Sends the component to the player
          *
