From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Wed, 10 Nov 2021 23:12:11 +0000
Subject: [PATCH] Bamboo Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 97e26767b9143e290fb232013090a91bb67bca7c..8522cec69b78befe6e3483034ffdd42e4e41c987 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1901,6 +1901,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Bamboo start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBambooConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bamboo end
+
         /**
          * Sends the component to the player
          *
