From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Wed, 10 Nov 2021 23:12:11 +0000
Subject: [PATCH] Bamboo Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1a38b9e616c6096d8d64c8abaa7178bc5f84e8d4..b5c3723f00177d6ce27be95162a97b8e682b4be9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1805,6 +1805,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Bamboo start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBambooConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bamboo end
+
         /**
          * Sends the component to the player
          *
