From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Wed, 10 Nov 2021 23:12:11 +0000
Subject: [PATCH] Bamboo Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2c4dfcf909c0214d25232a14635b96f92d4d974a..4f3406b85d4ef4d183db793409deceaccce89ed1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1793,6 +1793,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Bamboo start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getBambooConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Bamboo end
+
         /**
          * Sends the component to the player
          *
