From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Tue, 16 Nov 2021 01:25:38 +0000
Subject: [PATCH] Shearable Item Frames


diff --git a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
index e4c8b622cf3c7c57ada4c2e20a4b4a0f0b17de12..f962968235a3d36ca9645f48a0b7714c905c258f 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -195,6 +195,11 @@ public class ItemFrame extends HangingEntity {
                 // Paper end
                 this.dropItem(source.getEntity(), false);
                 this.playSound(this.getRemoveItemSound(), 1.0F, 1.0F);
+                // Bamboo start - Make visible if was sheared
+                if (this.isInvisible()) {
+                    this.setInvisible(false);
+                }
+                // Bamboo end
             }
 
             return true;
@@ -452,13 +457,20 @@ public class ItemFrame extends HangingEntity {
                     }
                 }
             } else {
-                // Paper start - call PlayerItemFrameChangeEvent
-                var event = new PlayerItemFrameChangeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.ItemFrame) this.getBukkitEntity(), this.getItem().asBukkitCopy(), PlayerItemFrameChangeEvent.ItemFrameChangeAction.ROTATE);
-                if (!event.callEvent()) return InteractionResult.FAIL;
-                setItem(ItemStack.fromBukkitCopy(event.getItemStack()), false, false);
-                // Paper end
-                this.playSound(this.getRotateItemSound(), 1.0F, 1.0F);
-                this.setRotation(this.getRotation() + 1);
+                // Bamboo start - Shearing makes item frame invisible
+                if (itemstack.is(Items.SHEARS) && player.getAbilities().mayBuild && !this.isInvisible() &&
+                    player.getBukkitEntity().hasPermission("bamboo.frames.shear")) {
+                    this.setInvisible(true);
+                } else {
+                    // Paper start - call PlayerItemFrameChangeEvent
+                    var event = new PlayerItemFrameChangeEvent((org.bukkit.entity.Player) player.getBukkitEntity(), (org.bukkit.entity.ItemFrame) this.getBukkitEntity(), this.getItem().asBukkitCopy(), PlayerItemFrameChangeEvent.ItemFrameChangeAction.ROTATE);
+                    if (!event.callEvent()) return InteractionResult.FAIL;
+                    setItem(ItemStack.fromBukkitCopy(event.getItemStack()), false, false);
+                    // Paper end
+                    this.playSound(this.getRotateItemSound(), 1.0F, 1.0F);
+                    this.setRotation(this.getRotation() + 1);
+                }
+                // Bamboo end
             }
 
             return InteractionResult.CONSUME;
