From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Sun, 21 Nov 2021 03:01:01 +0000
Subject: [PATCH] Disable Sheep Griefing Blocks


diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
index 16e66ae7cf761fabb74a035e2fd8c28615365dc2..9073c7e77314314d5a60848e8c731bcdd3101c0a 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/EatBlockGoal.java
@@ -3,7 +3,11 @@ package net.minecraft.world.entity.ai.goal;
 import java.util.EnumSet;
 import java.util.function.Predicate;
 import net.minecraft.core.BlockPos;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.Mob;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Block;
@@ -75,10 +79,16 @@ public class EatBlockGoal extends Goal {
             BlockPos blockposition = this.mob.blockPosition();
 
             if (EatBlockGoal.IS_TALL_GRASS.test(this.level.getBlockState(blockposition))) {
-                // CraftBukkit
-                if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition, Blocks.AIR.defaultBlockState(), !this.level.purpurConfig.sheepBypassMobGriefing && !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) { // Purpur
-                    this.level.destroyBlock(blockposition, false);
+                // Bamboo start
+                if (net.rocryn.bamboo.BambooConfig.sheepGriefGrass) {
+                    // CraftBukkit
+                    if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition, Blocks.AIR.defaultBlockState(), !this.level.purpurConfig.sheepBypassMobGriefing && !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) { // Purpur
+                        this.level.destroyBlock(blockposition, false);
+                    }
+                } else {
+                    this.level.playSound((Player) null, (Entity) this.mob, SoundEvents.GRASS_BREAK, SoundSource.BLOCKS, 1.0f, 1.0f);
                 }
+                // Bamboo stop
 
                 this.mob.ate();
                 this.mob.gameEvent(GameEvent.EAT, this.mob.eyeBlockPosition());
@@ -86,11 +96,17 @@ public class EatBlockGoal extends Goal {
                 BlockPos blockposition1 = blockposition.below();
 
                 if (this.level.getBlockState(blockposition1).is(Blocks.GRASS_BLOCK)) {
-                    // CraftBukkit
-                    if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition1, Blocks.AIR.defaultBlockState(),  !this.level.purpurConfig.sheepBypassMobGriefing && !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) { // Purpur
-                        this.level.levelEvent(2001, blockposition1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
-                        this.level.setBlock(blockposition1, Blocks.DIRT.defaultBlockState(), 2);
+                    // Bamboo start
+                    if (net.rocryn.bamboo.BambooConfig.sheepGriefGrass) {
+                        // CraftBukkit
+                        if (!CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockposition1, Blocks.AIR.defaultBlockState(), !this.level.purpurConfig.sheepBypassMobGriefing && !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) { // Purpur
+                            this.level.levelEvent(2001, blockposition1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
+                            this.level.setBlock(blockposition1, Blocks.DIRT.defaultBlockState(), 2);
+                        }
+                    } else {
+                        this.level.playSound((Player) null, (Entity) this.mob, SoundEvents.GRASS_BREAK, SoundSource.BLOCKS, 1.0f, 1.0f);
                     }
+                    // Bamboo stop
 
                     this.mob.ate();
                     this.mob.gameEvent(GameEvent.EAT, this.mob.eyeBlockPosition());
diff --git a/src/main/java/net/rocryn/bamboo/BambooConfig.java b/src/main/java/net/rocryn/bamboo/BambooConfig.java
index 6aeede351a536fffe3f4ab3e55f0a5f5100b728a..17d253f2859f733b66efff43a1f455298cccc13d 100644
--- a/src/main/java/net/rocryn/bamboo/BambooConfig.java
+++ b/src/main/java/net/rocryn/bamboo/BambooConfig.java
@@ -136,4 +136,7 @@ public class BambooConfig {
 
     public static Boolean preserveFuel = false;
     private static void preserveFuel() { preserveFuel = getBoolean("blocks.furnaces.preserve-fuel", preserveFuel); }
+
+    public static Boolean sheepGriefGrass = true;
+    private static void sheepGriefGrass() { sheepGriefGrass = getBoolean("mobs.passive.sheep.grief-grass", sheepGriefGrass); }
 }
