From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Fri, 17 Dec 2021 22:31:42 +0000
Subject: [PATCH] Compostable Poisonous Potato


diff --git a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
index 9315c819a0c113313a8e6f97e02b4d13c2cf62c0..0f496c4c513d3a07123fc85ac8dfe95d4daacbce 100644
--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
@@ -67,6 +67,7 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
         float f3 = 0.85F;
         float f4 = 1.0F;
 
+        ComposterBlock.add(0.1F, Items.POISONOUS_POTATO);
         ComposterBlock.add(0.3F, Items.JUNGLE_LEAVES);
         ComposterBlock.add(0.3F, Items.OAK_LEAVES);
         ComposterBlock.add(0.3F, Items.SPRUCE_LEAVES);
