From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Thu, 18 Nov 2021 04:06:18 +0000
Subject: [PATCH] Adjustable Monster Spawn Light Level


diff --git a/src/main/java/net/minecraft/world/entity/monster/Monster.java b/src/main/java/net/minecraft/world/entity/monster/Monster.java
index e94618f4c1552bb1ef20067aa3201c6574b5c307..92c40dc662eb601b5dcddb0d19244e2c83481d0a 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Monster.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Monster.java
@@ -23,6 +23,7 @@ import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.LightLayer;
 import net.minecraft.world.level.ServerLevelAccessor;
+import org.bukkit.Bukkit;
 
 public abstract class Monster extends PathfinderMob implements Enemy {
     public org.bukkit.craftbukkit.entity.CraftMonster getBukkitMonster() { return (org.bukkit.craftbukkit.entity.CraftMonster) super.getBukkitEntity(); } // Paper
@@ -93,12 +94,19 @@ public abstract class Monster extends PathfinderMob implements Enemy {
             return false;
         }
         // Purpur end
+
+        // Bamboo start
         if (world.getBrightness(LightLayer.SKY, pos) > random.nextInt(32)) {
             return false;
-        } else {
-            int i = world.getLevel().isThundering() ? world.getMaxLocalRawBrightness(pos, 10) : world.getMaxLocalRawBrightness(pos);
-            return i <= random.nextInt(8);
         }
+
+        if (world.getBrightness(LightLayer.BLOCK, pos) > net.rocryn.bamboo.BambooConfig.monsterSpawnLightLevel) {
+            return false;
+        }
+
+        int i = world.getLevel().isThundering() ? world.getMaxLocalRawBrightness(pos, 10) : world.getMaxLocalRawBrightness(pos);
+        return i <= random.nextInt(Math.max(net.rocryn.bamboo.BambooConfig.monsterSpawnLightLevel + 1, 8));
+        // Bamboo end
     }
 
     public static boolean checkMonsterSpawnRules(EntityType<? extends Monster> type, ServerLevelAccessor world, MobSpawnType spawnReason, BlockPos pos, Random random) {
