From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Tencryn <41876297+Tencryn@users.noreply.github.com>
Date: Tue, 28 Jun 2022 13:42:09 +0100
Subject: [PATCH] Creative One Punch Excludes Players


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 411593b1b105d62440d76b7bd1b8c74b701e3e75..ee2ed3338f0712b16c5e535f612879e008aa03b1 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2229,7 +2229,7 @@ public abstract class LivingEntity extends Entity {
 
            // Purpur start
            if (damagesource.getEntity() instanceof net.minecraft.world.entity.player.Player player && damagesource.getEntity().level.purpurConfig.creativeOnePunch) {
-               if (player.isCreative()) {
+               if (player.isCreative() && (!human || net.rocryn.bamboo.BambooConfig.creativeOnePunchExcludesPlayers)) { // Bamboo
                    double attackDamage = 0;
                    for (AttributeModifier modifier : player.getMainHandItem().getAttributeModifiers(EquipmentSlot.MAINHAND).get(Attributes.ATTACK_DAMAGE)) {
                        attackDamage += modifier.getAmount();
diff --git a/src/main/java/net/rocryn/bamboo/BambooConfig.java b/src/main/java/net/rocryn/bamboo/BambooConfig.java
index 7a0cedf4025cd71040ea68e63bc74062f8a01189..6c23f857452ebbfb0255b663de48acb07552a2a0 100644
--- a/src/main/java/net/rocryn/bamboo/BambooConfig.java
+++ b/src/main/java/net/rocryn/bamboo/BambooConfig.java
@@ -191,4 +191,7 @@ public class BambooConfig {
     private static void maxTridentLoyaltyEnchantment() { maxTridentLoyaltyEnchantment = getInt("max-enchantment-levels.trident-loyalty", maxTridentLoyaltyEnchantment); }
     private static void maxTridentRiptideEnchantment() { maxTridentRiptideEnchantment = getInt("max-enchantment-levels.trident-riptide", maxTridentRiptideEnchantment); }
     private static void maxWaterWalkerEnchantment() { maxWaterWalkerEnchantment = getInt("max-enchantment-levels.water-walker", maxWaterWalkerEnchantment); }
+
+    public static Boolean creativeOnePunchExcludesPlayers = false;
+    private static void creativeOnePunchExcludesPlayers() { creativeOnePunchExcludesPlayers = getBoolean("player.one-punch-in-creative-excludes-players", creativeOnePunchExcludesPlayers); }
 }
